TechnicalDocuments/Proposals/BabylonDesignLibrary.md

```diff
@@ -0,0 +1,196 @@
+# Babylon Design Library
+
+* Author(s): Giorgos Tsiapaliokas
+* Review Manager: Anders Ha
+
+## Introduction
+This proposal is about introducing a Design Library.
+
+## Motivation
+With our current infrastrure we have have reusable components which don't reuse the same style,
+we provide different styesheets based on the design.
+
+The idea behind this is to have reusable components with a predefined stylesheet displaying different data.
+The benefit of having a design library is
+1. to develop new UIs faster.
+2. to catch inconsistencies in the UI due to a deisgn flaw or due to an engineering mistake.
+
+The requirement is to cover most of the common use cases and *not* to extend to edge cases.
+Also we must reuse our current tools like `Bento`.
+
+For instance, in our codebase a common UI element is a rounded button with background color. 
+For that we use `Component.Button` and a stylesheet while this approach is functional is has a few drawbacks.
+The attributes of the button (background color, alpha, corner radius etc) can either 
+1. have been redefined in another part of the codebase which is good from a design point of view *and if* we search
+for them we can use them, *if we don't* search for them we will just reimplement something which already exists.
+2. or the attributes of the button haven't been redefined in another part of the application and we may not
+notice that so a design flaw will be implemented. We are humans after all :).
+
+
+## Proposed solution
+We can introduce `DesignLibrary` which will be responsible for creating these components following
+the semantics of the design styleguide which will be provided to us.
+
+The purpose of  `DesignLibrary` isn't to expose the full potential of `Bento` components but to 
+follow specific design guidelines.
+As an example of this is: 
+A `Bento` component supports different behaviours like  `Deletable` the goal of this
+proposal isn't to replicate them or to make a wrapper around them but to follow specific design guidelines
+and to offer reusable UI elements.
+
+Also with `DesignLibrary` we want to limit the exposure of specific components, in order to be able to introduce  
+drastic design changes in the future as easy as possible.
+For example `DesignLibrary.ComponentsBuilder.button` returns `AnyRenderable` instead of `Component.Button`
+because in the future we may decide that we need to use a different button component. We want to depend on abstractions.
+
+Furthermore, `DesignLibrary` can also host alerts, custom viewcontroller animations, etc.
+
+``` swift
+
+struct DesignLibrary {
+    let configuration: Configuration
+
+    var viewControllerBuilder: ViewControllerBuilder {
+        return ViewControllerBuilder(configuration: configuration)
```

Comment:
The `ViewControllerBuilder` might want to receive the whole `DesignLibrary` thing, e.g. a Bento based model.

Reply:
TBH I'm very skeptical about this dynamic theming conversations. One thing is to allow it to be configured at just at startup, another - to make it react just on system changes like change of content size category, and another thing is to make it completely dynamic in runtime, like switching to dark mode. Do we really need this? cc @RuiAAPeres

Reply:
This was actually an example, for now we won't have something like this.
But yes this is a valid comment, in the future if we want to add something like this 
then `ViewControllerBuilder.configuration` should be replaced with `ViewControllerBuilder.designLibrary`

Reply:
> Do we really need this?

Yes, at least for the new markets AFAIK.

Reply:
@andersio this does not answer what exactly out of three options I highlighted we need to think about. Having different styles for different partners is something we already have and it does not require any runtime updates. Or are we talking about changing how app looks like depending on what partner you register with?
Designing solution that will fit dark mode ATM is tricky I think as it's not very clear what API it will have (just a boolean in trait collections? notification? or the system colours will be inverted automatically somehow?). 
So I think we need to clarify what exactly we mean by "dynamic theming" to understand how dynamic our solution should be and to not overthink it.

Reply:
> Having different styles for different partners is something we already have and it does not require any runtime updates

The _plan_ for the new market, AFAICR, is having one app where themes change based on the associated partner of the user. So it is going to be different from static configurations we have per white label app.

> Designing solution that will fit dark mode ATM is tricky I think as it's not very clear what API it will have

Arguably speaking the API has already landed in `UITraitCollection` as `userInterfaceStyle`, albeit not being used for iOS at the moment. Since tvOS uses it for its dark mode, as an educated guess it is very unlikely that iOS diverges from this.

Reply:
Most importantly, regardless of needing backend driven themes or dark mode or both, building these on top of a design system while respecting accessibility settings are going to result in similar infrastructure given the dynamic factors involved, unless we are going after the path of "just kill the app to see your new theme".

Reply:
I agree that there should be some kind of observation of trait collections that should result in UI refresh, but I think weather it should be done through design system which should be made a `Property` or wether observation should be done separately and design system should just accept current traits as a parameter can be addressed later as such integration is a separate issue from having design library on its own.

Reply:
Dark theme and the ability to change it at runtime is still on the table. How this is to be achieved, that's a different matter.